quark/script/frida/__init__.py

```diff
@@ -0,0 +1,197 @@
+
+        :param targetMethod: the target API
+        :param methodParamTypes: the parameter types of the target API
+        :return: python list that holds calls to the target method
+        """
+        messageBuffer = []
+        methodId = self._getMethodIdentifier(targetMethod, methodParamTypes)
+
+        self.watchedMethods[methodId] = messageBuffer
+        self.script.exports.watch_method_call(targetMethod, methodParamTypes)
+
+        return messageBuffer
+
+    def stopWatchingMethodCall(
+        self, targetMethod: str, methodParamTypes: str
+    ) -> None:
+        """Stop tracking calls to the target method.
+
+        :param targetMethod: the target API
+        :param methodParamTypes: the parameter types of the target API
+        """
+        methodId = self._getMethodIdentifier(targetMethod, methodParamTypes)
+
+        if methodId in self.watchedMethods:
+            del self.watchedMethods[methodId]
+
+    def receiveMessage(self, messageFromFridaAgent: dict, _) -> None:
+        if messageFromFridaAgent["type"] == "error":
+            errorDescription = messageFromFridaAgent["description"]
+            print(errorDescription, file=sys.stderr)
+            return
+
+        receivedEvent = json.loads(messageFromFridaAgent["payload"])
+
+        eventType = receivedEvent.get("type", None)
+
+        if eventType == "CallCaptured":
+            methodId = tuple(receivedEvent["identifier"][0:2])
+
+            if methodId in self.watchedMethods:
+                messageBuffer = self.watchedMethods[methodId]
+                messageBuffer.append(receivedEvent)
+
+        elif eventType == "FailedToWatch":
+            methodId = tuple(receivedEvent["identifier"])
+            self.watchedMethods.pop(methodId)
+
+
+@functools.lru_cache
+def _setupFrida(
+    appPackageName: str, protocol="usb", **kwargs: Any
+) -> Tuple[Device, FridaSession, int]:
+    device = None
+    if protocol == "usb":
+        device = frida.get_usb_device(**kwargs)
+    elif protocol == "local":
+        device = frida.get_local_device(**kwargs)
+    elif protocol == "remote":
+        device = frida.get_remote_device(**kwargs)
+
+    processId = device.spawn([appPackageName])
+    session = device.attach(processId)
+
+    return device, session, processId
+
+
+@functools.lru_cache
+def _injectAgent(frida: FridaSession) -> MethodCallEventDispatcher:
+    dispatcher = MethodCallEventDispatcher(frida)
+
+    pathToFridaAgentSource = pkg_resources.resource_filename(
+        "quark.script.frida", "agent.js"
+    )
+
+    with open(pathToFridaAgentSource, "r") as fridaAgentSource:
+        fridaAgent = dispatcher.frida.create_script(fridaAgentSource.read())
+        fridaAgent.on("message", dispatcher.receiveMessage)
+        fridaAgent.load()
+        dispatcher.script = fridaAgent
+
+    return dispatcher
+
+
+@dataclass
+class Behavior:
+    _message: Dict[str, str]
+
+    def hasString(self, pattern: str, regex: bool = False) -> List[str]:
```

Comment:
Can we programmatically identify if the `pattern` is a regular expression?
If so, we can remove the parameter `regex`.